src/System.Linq/src/System/Linq/Take.SpeedOpt.cs

```diff
@@ -12,5 +13,58 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(count > 0);
+
+            if (source is IPartition<TSource> partition)
+            {
+                int length = partition.GetCount(true);
+                if (length > 0)
+                {
+                    return length - count > 0 ? partition.Skip(length - count) : partition;
+                }
+                else if (length == 0)
+                {
+                    return partition;
+                }
+            }
+            else if (source is IList<TSource> sourceList)
+            {
+                if (sourceList.Count > count)
+                {
+                    return new ListPartition<TSource>(sourceList, sourceList.Count - count, sourceList.Count);
+                }
+                else
+                {
+                    return new ListPartition<TSource>(sourceList, 0, sourceList.Count);
+                }
+            }
+            else if (source is ICollection<TSource> collection)
```

Comment:
This adds yet another interface type check onto the slow path.  It can also impact the limited interface slots issue that @AndyAyersMS was talking about.  Are there any other opterators where we have so many different conditionals?  I suggest we should stick with just IPartition, IList, and everything else, just as we have for Take.

Reply:
Also, looking at a code coverage report for this change, tests aren't exercising this code path, nor the readonly one below.

Reply:
> Are there any other opterators where we have so many different conditionals?

General Select<TSource, TResult>, Where<TSource> make for example three checks for class/interface. 

>Also, looking at a code coverage report for this change, tests aren't exercising this code path, nor the readonly one below.

The benchmarks are testing also for ICollection and it looks like code is faster by 25% in this case.

Reply:
> General Select<TSource, TResult>, Where make for example three checks for class/interface.

Thanks.  That's true, but Select is also one of if not the most commonly used methods in LINQ and deserves a bit more code to optimize it.  It's also not checking for `ICollection<T>`.

> The benchmarks are testing also for ICollection and it looks like code is faster by 25% in this case.

My point was that there are no unit tests exercising these code paths.

From a performance perspective, yes, adding a special case for ICollection will make ICollection go faster, at the expense of adding more code and working set for all cases, and penalizing non-collection inputs.

Reply:
> My point was that there are no unit tests exercising these code paths.

That's because unit tests doesn't test with ICollection not being IList, which is possible scenario. 

Nevertheless I've got rid of this ICollection stuff as you have suggested.